numpyro/contrib/autoguide/__init__.py

```diff
@@ -253,3 +255,83 @@ def quantiles(self, opt_state, quantiles):
                 transform = self._inv_transforms[name]
                 result.setdefault(name, []).append(transform(unconstrained_value))
         return result
+
+
+# TODO: remove when to_event is supported
+class _Normal(dist.Normal):
+    # work as Normal but has event_dim=1
+    def __init__(self, *args, **kwargs):
+        super(_Normal, self).__init__(*args, **kwargs)
+        self._event_shape = self._batch_shape[-1:]
+        self._batch_shape = self._batch_shape[:-1]
+
+    def log_prob(self, value):
+        return super(_Normal, self).log_prob(value).sum(-1)
+
+
+class AutoIAFNormal(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` uses a Diagonal Normal
+    distribution transformed via a
+    :class:`~numpyro.distributions.iaf.InverseAutoregressiveTransform`
+    to construct a guide over the entire latent space. The guide does not
+    depend on the model's ``*args, **kwargs``.
+
+    Usage::
+
+        guide = AutoIAFNormal(rng, model, get_params, hidden_dims=[20], skip_connections=True, ...)
+        svi_init, svi_update, _ = svi(model, guide, ...)
+    """
+    def __init__(self, rng, model, get_params_fn, prefix="auto", init_loc_fn=init_to_median,
+                 num_flows=3, **arn_kwargs):
+        self.arn_kwargs = arn_kwargs
+        self.arns = []
+        self.num_flows = num_flows
+        rng, *self.arn_rngs = random.split(rng, num_flows + 1)
+        super(AutoIAFNormal, self).__init__(rng, model, get_params_fn, prefix=prefix,
+                                            init_loc_fn=init_loc_fn)
+
+    def sample_posterior(self, rng, opt_state, *args, **kwargs):
+        sample_shape = kwargs.pop('sample_shape', ())
+        params = self.get_params(opt_state)
+        latent_size = np.size(self._init_latent)
+        flows = []
+        for i in range(self.num_flows):
+            arn_params = params['{}_arn__{}'.format(self.prefix, i)]
+            if i > 0:
+                flows.append(PermuteTransform(np.arange(latent_size)[::-1]))
+            flows.append(InverseAutoregressiveTransform(self.arns[i], arn_params))
+        iaf_dist = dist.TransformedDistribution(_Normal(np.zeros(latent_size), 1.), flows)
+        latent_sample = iaf_dist.sample(rng, sample_shape)
+        return self._unpack_latent(latent_sample)
+
+    def _sample_latent(self, *args, **kwargs):
```

Comment:
I am beginning to think that we will be better served by avoiding `*args, **kwargs` in these concrete classes, and just using fully qualified names like `sample_posterior(self, rng, opt_state, sample_shape=())`. That will both simplify the code and be self-documenting. Since I started with this, it is better to do it in a separate PR for both the classes.

Reply:
Yup, agree. :D